src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/pre_compute.cpp

```diff
@@ -0,0 +1,182 @@
+        const auto inverse_freq_fp32 = ov::as_type_ptr<ov::op::v0::Constant>(inverse_frequencies)->cast_vector<float>();
+        size_t rotary_ndims = ov::shape_size(inverse_frequencies->get_shape()) * 2;
+
+        std::vector<ov::float16> lut_sin(max_position_embeddings * rotary_ndims, 0.0f);
+        std::vector<ov::float16> lut_cos(max_position_embeddings * rotary_ndims, 0.0f);
+
+        // rotate_half style cos/sin table:
+        //   y1 = cos(m*xita_i) * x1 - sin(m*xita_i) * x2
+        //   y2 = cos(m*xita_i) * x2 + sin(m*xita_i) * x1
+        //
+        for (size_t i = 0, k = 0; i < rotary_ndims; i += 2, k++) {
+            //auto xita_i = 1.0 / std::pow(10000.0, static_cast<double>(i) / rotary_ndims);
+            auto xita_i = inverse_freq_fp32[i >> 1];
+            ov::float16* psin = lut_sin.data();
+            ov::float16* pcos = lut_cos.data();
+            for (size_t m = 0; m < max_position_embeddings; m++, psin += rotary_ndims, pcos += rotary_ndims) {
+                auto vsin = std::sin(xita_i * m);
+                auto vcos = std::cos(xita_i * m);
+                pcos[k] = ov::float16{vcos};
+                pcos[k + rotary_ndims / 2] = pcos[k];
+
+                psin[k] = ov::float16{vsin};
+                psin[k + rotary_ndims / 2] = psin[k];
+            }
+        }
+        auto Cos = makeConst(ov::element::f16, ov::Shape({1, max_position_embeddings, rotary_ndims}), lut_cos);
+        auto Sin = makeConst(ov::element::f16, ov::Shape({1, max_position_embeddings, rotary_ndims}), lut_sin);
+
+        return {Cos, Sin};
+    }
+}  // namespace
+
+ov::npuw::patterns::pre_compute::RopePatternLLama2::RopePatternLLama2() {
+    auto shape_of = opp::wrap_type<ov::op::v3::ShapeOf>({opp::any_input()});
+    auto gather = opp::wrap_type<ov::op::v8::Gather>({shape_of, opp::any_input(), opp::any_input()});
+    auto concat_1 = opp::wrap_type<ov::op::v0::Concat>({gather, opp::any_input(), opp::any_input()});
+    //here we can seen inverse frequencies as a parameter or constant depending on partitioner passes
+    auto inv_freq = opp::wrap_type<ov::op::v0::Constant>();
+    auto inv_freq_convert = opp::optional<ov::op::v0::Convert>({inv_freq->output(0)});
+    auto broadcast = opp::wrap_type<ov::op::v3::Broadcast>({inv_freq_convert, concat_1});
+    auto position_ids = opp::wrap_type<ov::op::v0::Parameter>();
+    auto unsqueeze = opp::wrap_type<ov::op::v0::Unsqueeze>({position_ids, opp::wrap_type<ov::op::v0::Constant>()});
+    auto convert = opp::wrap_type<ov::op::v0::Convert>({unsqueeze});
+    auto matmul = opp::wrap_type<ov::op::v0::MatMul>({broadcast, convert});
+    auto transpose = opp::wrap_type<ov::op::v1::Transpose>({matmul, opp::any_input()});
+    auto concat_2 = opp::wrap_type<ov::op::v0::Concat>({transpose, opp::any_input()});
+    auto sin_cos = opp::wrap_type<ov::op::v0::Sin, ov::op::v0::Cos>({concat_2});
```

Comment:
this is like OR, isn't it? I see you check in the callback that there should be two exact Sin/Cos operations, wouldn't it help to explicitly put `sin` _and_ `cos`? Otherwise, will your pattern get matched twice?